Fallback to :memory: on repeated failure to load history

minrk opened the pull request:
two consecutive failures to load history will result in falling back on :memory:

I also fixed a case where multiple consecutive failures to load would erase the original db with empty ones created after the first failure. The date is now in the corrupt-db filename.

closes #9412

minrk committed:
Fallback to :memory: on repeated failure to load history

two consecutive failures to load history will result
in falling back on :memory:

takluyver commented on IPython/core/history.py:
```diff
@@ -81,18 +80,29 @@ def catch_corrupt_db(f, self, *a, **kw):
+                # Try to move the file out of the way
+                base,ext = os.path.splitext(self.hist_file)
+                now = datetime.datetime.now().isoformat().replace(':', '.')
+                newpath = base + '-corrupt-' + now + ext
```
I appreciate the reason for this, but I'm not sure I like the solution - in a case like #9412, if the user doesn't configure a different location, it's presumably going to create two new, empty, corrupt database files for each IPython session, until the profile directory is littered with thousands of them.

Maybe it could just include the corrupt_db_counter - that way you preserve all the files from the last attempt to start IPython, but they're not accumulating.

minrk replied:
That would mean starting IPython twice wipes out your history. In the NFS case prompting this, the db is not actually bad, SQLite is just failing to open it due to locking problems. So we should be making sure that we never overwrite the 'good' db, otherwise the '-corrupt' backup is pretty pointless.

takluyver replied:
Right, but in the NFS case prompting this, it's failing to even create the database. So there's never a useful 'good' database worth preserving.

I'm not too concerned if starting IPython twice with a corrupt database wipes out your history - if we can't use that database, your history is effectively gone anyway, and I think ~no-one is going to take the time to recover it. I'd be happy if it kept it for a bit longer, or was smart enough not to move a database that it had just tried and failed to create. I'm just not happy about the ease with which this could create a great many useless files.

minrk replied:
Updated to limit the no-clobber logic to files >= 16k, which should mean they have *some* content, at least. That means the thrashing where it never works should create at most one file.

minrk committed:
only avoid clobbering corrupt dbs that have some content

this logic can be triggered by connection failures that do not mean the file is corrupt,
so we don't want to lose data unnecessarily.

takluyver commented:
Thanks. I'm happy with this now. Leaving it for a day or so in case other people want to look at it.

minrk commented:
Thanks for the review!

Carreau commented:
Thanks for that!
